src/moq_instance.cpp

```diff
@@ -0,0 +1,1672 @@
+            send_announce(conn_it->second, track_delegate->getTrackNamespace());
+
+        } else {
+            auto pub_n_it = pub_ns_it->second.find(th.track_name_hash);
+            if (pub_n_it == pub_ns_it->second.end()) {
+                _logger->info << "Publish track has new track "
+                              << " namespace hash: " << th.track_namespace_hash
+                              << " name hash: " << th.track_name_hash
+                              << std::flush;
+            }
+        }
+
+        // Set the track delegate for pub/sub
+        conn_it->second.pub_tracks_by_name[th.track_namespace_hash][th.track_name_hash] = track_delegate;
+
+        track_delegate->_mi_conn_id = conn_id;
+        track_delegate->_mi_send_data_ctx_id = _transport->createDataContext(
+          conn_id,
+          track_delegate->_mi_track_mode == MoQTrackDelegate::TrackMode::DATAGRAM ? false : true,
+          track_delegate->_def_priority,
+          false);
+
+#ifndef LIBQUICR_WITHOUT_INFLUXDB
+        Name n;
+        n += th.track_fullname_hash;
+
+        _mexport.set_data_ctx_info(conn_id, track_delegate->_mi_send_data_ctx_id,
+                                   {.subscribe = false, .nspace = {n, 64} } );
+#endif
+
+
+        // Setup the function for the track delegate to use to send objects with thread safety
+        track_delegate->_mi_sendObjFunc =
+          [&,
+           track_delegate = track_delegate,
+           subscribe_id = track_delegate->getSubscribeId()](uint8_t priority,
+                                                            uint32_t ttl,
+                                                            bool stream_header_needed,
+                                                            uint64_t group_id,
+                                                            uint64_t object_id,
+                                                            std::span<const uint8_t> data) -> MoQTrackDelegate::SendError {
+            return send_object(track_delegate,
+                               priority,
+                               ttl,
+                               stream_header_needed,
+                               group_id,
+                               object_id,
+                               data);
+        };
+
+        return th.track_fullname_hash;
+    }
+
+    MoQTrackDelegate::SendError MoQInstance::send_object(std::weak_ptr<MoQTrackDelegate> track_delegate,
+                                                         uint8_t priority,
+                                                         uint32_t ttl,
+                                                         bool stream_header_needed,
+                                                         uint64_t group_id,
+                                                         uint64_t object_id,
+                                                         std::span<const uint8_t> data)
+    {
+
+        auto td = track_delegate.lock();
```

Comment:
Should we check this isn't null, just in case?

Reply:
To that end, should we just pass a shared_ptr? Or do we not want it to be kept alive by this method?

Reply:
I think locking at the beginning and trusting it isn't nullptr is equivalent to just passing the shared_ptr by value. If we don't want the copy of the shared_ptr to share ownership, then maybe we should just pass a shared_ptr reference?

Reply:
Moq instance, at least right now, controls setting this from a shared pointer, which is guaranteed to not be null.  In this sense it's known and doesn't need to be checked if null, but I suppose it wouldn't impact it if we checked if it was null or not, but it would be error if it was nullptr, which probably should result in an abort.